### PR TITLE
config: improve error messages for extra & missing fields

### PR DESCRIPTION
**Pull request by cjp256**
Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

### CODE REVIEW

File: charmcraft/config.py [hunk 1]
```diff
@@ -150,7 +171,18 @@ def format_pydantic_errors(errors):
...
-        combined.append(f"- {formatted_msg} in field {formatted_loc!r}")
+        if formatted_msg == "field required":
+            field_name, location = printable_field_location_split(formatted_loc)
+            combined.append(
+                f"- field {field_name} required for {location} configuration"
+            )
+        elif formatted_msg == "extra fields not permitted":
+            field_name, location = printable_field_location_split(formatted_loc)
+            combined.append(
+                f"- extra field {field_name} not permitted for {location} configuration"
```
Comment (facundobatista) [inline charmcraft/config.py:182]: "required for" and "permitted for", or (taking in consideration that the problematic field happens inside a config key, "required in" and "permitted in"?

**Review by facundobatista — APPROVED** [review]
+1, but please take a look to the question inside. Thanks!

File: charmcraft/config.py (hunk 1 above)
Comment (cjp256) [reply charmcraft/config.py:182]: I actually did it both ways and kinda flip-flopped.  Take your pick and I'll update it if desired :)

File: charmcraft/config.py [hunk 2]
```diff
@@ -150,7 +171,18 @@ def format_pydantic_errors(errors):
         formatted_loc = format_pydantic_error_location(error["loc"])
         formatted_msg = format_pydantic_error_message(error["msg"])
 
-        combined.append(f"- {formatted_msg} in field {formatted_loc!r}")
+        if formatted_msg == "field required":
```
Comment (cmatsuoka) [inline charmcraft/config.py:174]: Could this be affected by locale?

File: charmcraft/config.py [hunk 3]
```diff
@@ -150,7 +171,18 @@ def format_pydantic_errors(errors):
...
+            combined.append(
+                f"- field {field_name} required in {location} configuration"
+            )
+        elif formatted_msg == "extra fields not permitted":
+            field_name, location = printable_field_location_split(formatted_loc)
+            combined.append(
+                f"- extra field {field_name} not permitted in {location} configuration"
+            )
+        else:
+            combined.append(f"- {formatted_msg} in field {formatted_loc!r}")
```
Comment (cmatsuoka) [inline charmcraft/config.py:185]: This kind of handling would be useful in craft-parts too. Perhaps we could move it to an external common toolbox later.

**Review by cmatsuoka — APPROVED** [review]

File: charmcraft/config.py (hunk 2 above)
Comment (cjp256) [reply charmcraft/config.py:174]: negative, but I wonder if we can catch / determine the error class directly, that may be better in the future

File: charmcraft/config.py (hunk 3 above)
Comment (cjp256) [reply charmcraft/config.py:185]: Sounds like a good idea

**Pull request merged**